### PR TITLE
Add support for managed identity for redis

### PR DESCRIPTION
**Pull request by twsouthwick**
This change does the following:

- Adds a new application module for the redis resource which handles adding the passwordless connection string to the app configuration
- Updates to a newer redis bicep module that allows using entra
- Adds a parameter to add entra users to the redis service
- Refactors the registration of the redis distributed cache to use managed identity

### CODE REVIEW

File: infra/main.bicep [hunk 1]
```diff
@@ -442,6 +442,7 @@ module application './modules/application-resources.bicep' = {
     frontDoorSettings: frontdoor.outputs.settings
 
     // Settings
+    principalId: principalId
```
Comment (KSchlobohm) [inline infra/main.bicep:445]: can you remove this by using the `principalId` in the DeploymentSettings object?

Comment (twsouthwick) [reply infra/main.bicep:445]: for sure - didn't realize it was there!

File: infra/core/database/azure-cache-for-redis.bicep [hunk 2]
```diff
@@ -46,6 +46,16 @@ type PrivateEndpointSettings = {
...
 
+// From: infra/types/RedisUser.bicep
+@description('Type describing the user for redis.')
+type RedisUser = {
+  @description('The object id of the user.')
+  objectId: string
+
+  @description('The alias of the user')
+  alias: string
+}
```
Comment (adrianhall) [inline infra/core/database/azure-cache-for-redis.bicep:57]: You probably want to add accessPolicyName here - that way your core module can assign each user to a different access policy if needed.   Add

accessPolicyName: 'Data Owner' | 'Data Contributor' | 'Data Reader'

File: infra/core/database/azure-cache-for-redis.bicep [hunk 3]
```diff
@@ -105,11 +115,21 @@ param redisCacheCapacity int = 1
...
+@description('Specify name of Built-In access policy to use as assignment.')
+@allowed([
+  'Data Owner'
+  'Data Contributor'
+  'Data Reader'
+])
+param builtInAccessPolicyName string = 'Data Reader'
+
+param users RedisUser[] = []
+
```
Comment (adrianhall) [inline infra/core/database/azure-cache-for-redis.bicep:127]: Combine the accessPolicyName into the RedisUser object.

File: infra/core/database/azure-cache-for-redis.bicep [hunk 4]
```diff
@@ -121,9 +141,23 @@ resource cache 'Microsoft.Cache/redis@2023-04-01' = {
...
+    }
   }
 }
 
+@batchSize(1)
+resource redisCacheBuiltInAccessPolicyAssignment 'Microsoft.Cache/redis/accessPolicyAssignments@2023-08-01' = [for user in users: {
+  name: guid(resourceGroup().id, name, builtInAccessPolicyName, user.objectId)
+  parent: cache
+  properties: {
+    accessPolicyName: builtInAccessPolicyName
```
Comment (adrianhall) [inline infra/core/database/azure-cache-for-redis.bicep:155]: This is where you would do the user.accessPolicyName to get it added to the right access policy.

File: infra/modules/application-resources.bicep [hunk 5]
```diff
@@ -452,25 +452,33 @@ module webFrontendFrontDoorRoute '../core/security/front-door-route.bicep' = if
...
-      name: resourceNames.redisPrivateEndpoint
-      resourceGroupName: resourceNames.spokeResourceGroup
-      subnetId: subnets[resourceNames.spokePrivateEndpointSubnet].id
-    } : null
+    subnets: subnets
+    tags: moduleTags
+    users: deploymentSettings.principalId == null ? [
+      {
+       alias: appManagedIdentity.name
+       objectId: appManagedIdentity.outputs.principal_id
```
Comment (adrianhall) [inline infra/modules/application-resources.bicep:470]: ... and just add accessPolicyName: 'Data Contributor' here.

File: infra/types/RedisUser.bicep [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// From: infra/types/RedisUser.bicep
+@description('Type describing the user for redis.')
+type RedisUser = {
+  @description('The object id of the user.')
+  objectId: string
+
+  @description('The alias of the user')
+  alias: string
```
Comment (adrianhall) [inline infra/types/RedisUser.bicep:8]: Ditto the change we made earlier.

**Review by adrianhall — APPROVED** [review]
My comments make the whole thing "better", but are in no way blocking.  Looks good otherwise.

File: infra/core/database/azure-cache-for-redis.bicep (hunk 3 above)
Comment (twsouthwick) [reply infra/core/database/azure-cache-for-redis.bicep:127]: i did that but hadn't pushed that change yet (still figuring out how to type things in bicep). I'll include it in this pr

**Review by adrianhall — APPROVED** [review]

**Comment by twsouthwick** [comment]
I validated that this doesn't need any docs changes (couldn't find reference to the password/connection string/etc) and the demo continues to work as documented.

This should close https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1881951

**Comment by KSchlobohm** [comment]
@twsouthwick 

It looks like this PR modifies the `Startup.cs` that is presented in the docs. Specifically, the changes regarding the new `TokenCredential` class that change the following sections.

* https://learn.microsoft.com/en-us/azure/architecture/web-apps/guides/reliable-web-app/dotnet/apply-pattern#use-defaultazurecredential-to-set-up-code
* https://learn.microsoft.com/en-us/azure/architecture/web-apps/guides/reliable-web-app/dotnet/apply-pattern#use-the-cache-aside-pattern

**Comment by twsouthwick** [comment]
Thanks @KSchlobohm - missed that part of the docs. I created a branch here https://github.com/twsouthwick/architecture-center/tree/rwa-servicing to update docs and will open a pr there when we're ready to merge here

**Pull request merged**